simulator/mips/t/unit_test.cpp

```diff
@@ -236,10 +171,621 @@ TEST_CASE( "MIPS32_instr_disasm: Process_Disasm_Branches")
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 0xffffffff, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+}
+
+TEST_CASE ( "MIPS32_instr: andi")
+{
+    CHECK(MIPS32Instr(0x313104d2).get_disasm() == "andi $s1, $t1, 0x4d2");
+    MIPS32Instr instr( "andi");
+
+    instr.set_v_src( 0, 0);
+    instr.set_v_imm( 0);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_imm( 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_imm( 0xffff);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
+    instr.set_v_src( 0xffffffff, 0);
+    instr.set_v_imm( 0xffff);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x0000ffff);
+}
+
+TEST_CASE( "MIPS32_instr: beq")
+{
+    CHECK(MIPS32Instr(0x1229ffff).get_disasm() == "beq $s1, $t1, -1");
+    CHECK(MIPS32Instr(0x1229000e).get_disasm() == "beq $s1, $t1, 14");
+    MIPS32Instr instr( "beq");
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( 0, 0);
+    instr.set_v_src( 0, 1);
+    instr.set_v_imm( 1);
+    instr.execute();
+    CHECK( instr.get_new_PC() == instr.get_PC() + 8);
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( 0xffffffff, 0);
+    instr.set_v_src( 0xffffffff, 1);
+    instr.set_v_imm( 1);
+    instr.execute();
+    CHECK( instr.get_new_PC() == instr.get_PC() + 8);
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( 0, 0);
+    instr.set_v_src( 1, 1);
+    instr.execute();
+    CHECK( instr.get_new_PC() == instr.get_PC() + 4);
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( 0xffffffff, 0);
+    instr.set_v_src( 0xffffffff, 1);
+    instr.set_v_imm( 1024);
+    instr.execute();
+    CHECK( instr.get_new_PC() == instr.get_PC() + 4 + 1024 * 4);
+
+    instr.set_new_PC( instr.get_PC() + 4);
+    instr.set_v_src( 0, 0);
+    instr.set_v_src( 0, 1);
+    instr.set_v_imm( 0xffffffff - 1024 + 1);
+    instr.execute();
+    CHECK( instr.get_new_PC() == instr.get_PC() + 4 - 1024 * 4);
+}
+
+TEST_CASE( "MIPS32_instr: break")
+{
+    CHECK(MIPS32Instr(0x0000000d).get_disasm() == "break");
+    MIPS32Instr instr( "break");
+    instr.execute();
+    CHECK( instr.trap_type() == Trap::BREAKPOINT);
+}
+
+TEST_CASE( "MIPS64_instr: dsllv")
+{
+    CHECK(MIPS32Instr(0x03298814).get_disasm() == "dsllv $s1, $t1, $t9");
+    MIPS64Instr instr( "dsllv");
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 0, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 1, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 2);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 32, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 0x100000000);
+
+    instr.set_v_src( 1, 0);
+    instr.set_v_src( 64, 1);
+    instr.execute();
+    CHECK( instr.get_v_dst() == 1);
```

Comment:
Ooops, looks like we have a real bug here!
Please comment that test case and open an issue.